wire-runtime-swift/src/main/swift/ExtensibleUnknownFields.swift

```diff
@@ -0,0 +1,153 @@
+        guard let data = rawData[fieldNumber] else {
+            return nil
+        }
+        return try memoizedData[fieldNumber]?.get {
+            try compute(data)
+        }
+    }
+
+    mutating func setParsedField<T>(fieldNumber: Key, value: T?, compute: (T) throws -> Value) rethrows {
+        guard let value else {
+            rawData[fieldNumber] = nil
+            memoizedData[fieldNumber] = nil
+            return
+        }
+        let data = try compute(value)
+        rawData[fieldNumber] = data
+        memoizedData[fieldNumber] = MemoizedValue(cached: value)
+    }
+}
+
+extension ExtensibleUnknownFields: Codable, Equatable, Hashable, Sendable {
+    public init(from decoder: any Decoder) throws {
+        let rawData = try BackingStore(from: decoder)
+        self.init(rawData: rawData)
+    }
+
+    public func encode(to encoder: any Encoder) throws {
+        try rawData.encode(to: encoder)
+    }
+
+    public static func == (lhs: Self, rhs: Self) -> Bool {
+        lhs.rawData == rhs.rawData
+    }
+
+    public func hash(into hasher: inout Hasher) {
+        rawData.hash(into: &hasher)
+    }
+}
+
+extension ExtensibleUnknownFields: CustomDebugStringConvertible, CustomStringConvertible {
+    public var description: String { rawData.description }
+    public var debugDescription: String { rawData.debugDescription }
+}
+
+extension ExtensibleUnknownFields: ExpressibleByDictionaryLiteral {
+    public init(dictionaryLiteral: (Key, Value)...) {
+        self.init(uniqueKeysWithValues: dictionaryLiteral)
+    }
+
+    public init<S>(uniqueKeysWithValues keysAndValues: S) where S : Sequence, S.Element == (Key, Value) {
+        self.init(rawData: BackingStore(uniqueKeysWithValues: keysAndValues))
+    }
+}
+
+extension ExtensibleUnknownFields: Collection {
+    public typealias Index = BackingStore.Index
+    public typealias Element = BackingStore.Element
+
+    public var isEmpty: Bool { rawData.isEmpty }
+    public var count: Int { rawData.count }
+
+    public var startIndex: BackingStore.Index {rawData.startIndex }
+    public var endIndex: BackingStore.Index { rawData.endIndex }
+    public var indices: BackingStore.Indices { rawData.indices }
+    public var underestimatedCount: Int { rawData.underestimatedCount }
+
+    public func index(after i: BackingStore.Index) -> BackingStore.Index {
+        rawData.index(after: i)
+    }
+
+    public subscript(index: Index) -> Element {
+        rawData[index]
+    }
+
+    public func index(forKey key: Key) -> Index? {
+        rawData.index(forKey: key)
+    }
+
+    public var keys: BackingStore.Keys { rawData.keys }
+    public var values: BackingStore.Values { rawData.values }
+}
+
+private class MemoizedValue: @unchecked Sendable {
+    private var cached: Any?
+
+    private let lock = NSLock()
+
+    init(cached: Any? = nil) {
+        self.cached = cached
+    }
+
+    func get<T>(compute: () throws -> T) rethrows -> T {
```

Comment:
- [ ] DocC

```swift
/// This will attempt to fetch a cached value in the expected output type.
///
/// If the requested value is the wrong type _or_ of it has not been instantiated yet, it will be computed.
```